[JSC] Syntax error should be thrown for logical assignment expressions where the LHS is a function call.

sosukesuzuki opened the pull request:
#### b8cb1ca3b5f0debf343a81accdd77cffe07622f5
<pre>
[JSC] Syntax error should be thrown for logical assignment expressions where the LHS is a function call.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272524">https://bugs.webkit.org/show_bug.cgi?id=272524</a>

Reviewed by Alexey Shvayka.

JSC throws a Reference Error instead of a Syntax Error for logical assignment expressions where the LHS is a function call in sloppy mode. According to the commit[2] when logical assignment expressions were implemented in JSC, this was intentional. However, four years have passed, and the related specification has not been updated, while other JS engines throw a Syntax Error.

This patch changes the behavior to throw a Syntax Error for such logical assignment expressions even in sloppy mode.

[1]: <a href="https://tc39.es/ecma262/#sec-assignment-operators-static-semantics-early-errors">https://tc39.es/ecma262/#sec-assignment-operators-static-semantics-early-errors</a>
[2]: <a href="https://commits.webkit.org/223412@main">https://commits.webkit.org/223412@main</a>

* JSTests/stress/bytecode-for-rmw-with-invalid-right-side.js:
* JSTests/stress/logical-assignment-with-func-call-lhs.js: Added.
(shouldThrowSyntaxError):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::isSimpleAssignmentTarget):
(JSC::Parser&lt;LexerType&gt;::parseAssignmentExpression):
* Source/JavaScriptCore/parser/Parser.h:

Canonical link: <a href="https://commits.webkit.org/277536@main">https://commits.webkit.org/277536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2635096a45ad5ff44753a96610e64413b9afaa2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26724 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50223 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43589 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32426 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38708 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48123 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24323 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40970 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20015 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21786 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42145 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5583 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/40821 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43869 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52103 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47031 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22575 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18905 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46017 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23848 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45047 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10566 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24636 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54531 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23567 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11176 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (approved):
Nice, thank you! We might want to adjust / remove the failing test.

Commit:
[JSC] Syntax error should be thrown for logical assignment expressions where the LHS is a function call.
https://bugs.webkit.org/show_bug.cgi?id=272524

Reviewed by Alexey Shvayka.

JSC throws a Reference Error instead of a Syntax Error for logical assignment expressions where the LHS is a function call in sloppy mode. According to the commit[2] when logical assignment expressions were implemented in JSC, this was intentional. However, four years have passed, and the related specification has not been updated, while other JS engines throw a Syntax Error.

This patch changes the behavior to throw a Syntax Error for such logical assignment expressions even in sloppy mode.

[1]: https://tc39.es/ecma262/#sec-assignment-operators-static-semantics-early-errors
[2]: https://commits.webkit.org/223412@main

* JSTests/stress/bytecode-for-rmw-with-invalid-right-side.js:
* JSTests/stress/logical-assignment-with-func-call-lhs.js: Added.
(shouldThrowSyntaxError):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::isSimpleAssignmentTarget):
(JSC::Parser<LexerType>::parseAssignmentExpression):
* Source/JavaScriptCore/parser/Parser.h:

Canonical link: https://commits.webkit.org/277536@main
